### PR TITLE
07748 Postconsensus signature gathering

### LINKED ISSUE #7748: Implement more robust round signing
Consider both pre and post consensus transactions for state signature collection instead of just pre-consensus.

### PR DESCRIPTION
**Pull request by alittley**
- closes #7748

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedStateManager.java [hunk 1]
```diff
@@ -304,6 +304,31 @@ public synchronized void handlePreconsensusSignatureTransaction(
...
+     */
+    public synchronized void handlePostconsensusSignatureTransaction(
+            @NonNull NodeId signerId, @NonNull StateSignatureTransaction transaction) {
+
+        Objects.requireNonNull(signerId);
+
+        final long round = transaction.getRound();
+
+        try (final ReservedSignedState reservedState = getIncompleteState(round)) {
+            // it isn't possible to receive a post-consensus signature transaction for a future round,
```
Comment (poulok) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedStateManager.java:321]: It is possible if this node is slow and has not yet reached consensus on a round that other nodes have. We should store signatures for a certain number of rounds in the future and only discard these if they are too far ahead. When a new state is created by this node, apply any signatures for it that were previously collected.

Comment (alittley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedStateManager.java:321]: @poulok If the signature transaction for a previous round is being handled by `nodeX`, isn't it guaranteed that the round the signature is on has _also_ already come to consensus for `nodeX`?

Comment (poulok) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedStateManager.java:321]: ahh, yes of course, you are correct.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedStateManager.java [hunk 2]
```diff
@@ -304,6 +304,31 @@ public synchronized void handlePreconsensusSignatureTransaction(
...
     }
 
+    /**
+     * An observer of post-consensus state signatures.
+     *
+     * @param signerId    the node that created the signature
+     * @param transaction the signature transaction
+     */
+    public synchronized void handlePostconsensusSignatureTransaction(
+            @NonNull NodeId signerId, @NonNull StateSignatureTransaction transaction) {
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedStateManager.java:314]:
```suggestion
    public synchronized void handlePostconsensusSignatureTransaction(
            @NonNull final NodeId signerId, @NonNull final StateSignatureTransaction transaction) {
```

Comment (alittley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedStateManager.java]: fixed

**Comment by lpetrovic05** [comment]
i dont understand the reasoning behind this change

**Comment by cody-littley** [comment]
@lpetrovic05 
> i dont understand the reasoning behind this change

When nodes get slightly out of sync with each other, sometimes we end up discarding signature transactions. This happens because we set a limit for the number of rounds in the future that we will accept signature transactions. This limit is necessary to prevent a denial of service attack.

By also ingesting the signatures after events reach consensus, we allow nodes that are behind their peers to collect signatures they may have discarded at the preconsensus phase. We still can't provide a 100% guarantee we will get all signatures (events that go stale may cause lost signatures), but it closes a major signature "leak".

Most often we observe this problem in JRS tests. For whatever reason, the test runner sometimes starts nodes in a why that causes them to be out of sync near the beginning of the test. The primary goal of this change was to make tests less flaky... although it's a change we would want to make anyways, even if the tests weren't flaking.

**Pull request merged**